crates/miden-lib/asm/miden/contracts/auth/multisig.masm

```diff
@@ -1,24 +1,273 @@
+    # => [index, MSG]
+
+    exec.account::get_map_item
+    # => [LOOKUP_VAL]
+
+    padw
+    # => [EMPTY_WORD, LOOKUP_VAL]
+
+    assert_eqw
+    # => []
+
+    # Add transaction to executed transactions map.
+    # ---------------------------------------------------------------------------------------------
+
+    push.1.1.1.1
+
+    padw loc_loadw.0
+
+    push.EXECUTED_TXS_SLOT
+    # => [index, KEY, VALUE]
+
+    # Set the key value pair in the map to mark transaction as executed.
+    exec.account::set_map_item
+    # => [OLD_MAP_ROOT, OLD_MAP_VALUE]
+
+    dropw dropw
+    # => []
+end
+
+#! Verify signatures for all required signers in a loop.
+#!
+#! This procedure iterates through the required number of signers, fetches their public keys
+#! from the account storage map, and verifies their signatures against the transaction message.
+#! If any required signature is missing, it emits an UNAUTHORIZED_EVENT.
+#!
+#! Inputs:  [num_required_signatures]
+#! Outputs: []
+proc.verify_signatures_loop
+    # => [num_required_signatures]
+
+    dup mem_store.THRESHOLD
+    # => [num_required_signatures]
+
+    padw mem_loadw.MESSAGE movup.4
+    # => [num_required_signatures, MESSAGE]
+
+    # Counter `i` starts at `num_required_signatures` and counts down to 0
+    # => [i, MESSAGE]
+
+    # Loop through required signers and verify signatures.
+    # ---------------------------------------------------------------------------------------------
+
+    dup neq.0
+    while.true
+        # => [i, MESSAGE]
+
+        # Fetch public key from storage map.
+        # -----------------------------------------------------------------------------------------
+
+        dup sub.1 push.0.0.0 push.PUBLIC_KEYS_MAP_SLOT
+        # => [owner_key_slot, [0, 0, 0, i-1], i, MESSAGE]
+
+        exec.account::get_map_item
+        # => [OWNER_PUB_KEY, i, MESSAGE]
+
+        mem_storew.TEMP_PK
+        # => [OWNER_PUB_KEY, i, MESSAGE]
+
+        # Check if signature exists for this signer.
+        # -----------------------------------------------------------------------------------------
+
+        movup.4 movdn.8 swapw dupw movdnw.2
+        # => [MESSAGE, OWNER_PUB_KEY, MESSAGE, i]
+
+        hmerge
+        # => [SIG_KEY, MESSAGE, i]
+
+        adv.has_mapkey
+        # => [SIG_KEY, MESSAGE, i]
+
+        adv_push.1
+        # => [has_signature, SIG_KEY, MESSAGE, i]
+
+        # if h(msg, pk) => msg exists in AdviceMap check the signature
+        if.true
+            dropw dupw
+            # => [MSG, MSG, i]
+
+            # Verify the signature against the public key and message.
+            # -----------------------------------------------------------------------------------------
+
+            padw mem_loadw.TEMP_PK
+            # => [PK, MSG, MSG, i]
+
+            swapw
+            # => [MSG, PK, MSG, i]
+
+            # Emit the authentication request event that pushes a signature for the message to the advice stack.
+            emit.AUTH_REQUEST
```

Comment:
As per my understanding in basic wallet keystore is present at site of execution, so when this event is emitted, signature is provided instantly by the keystore. But in the case of mutlisig, approver keystore is not present at site of execution so even after emitting this event we won't be able to get the signature.

Reply:
This procedure will be invoked twice:
1. First, dry-run. We don't have enough signatures, so we will hit `emit.UNAUTHORIZED` (which is our definition of dry run for this purpose). This is the _Initiator_ invoking the procedure.
2. Second, once all the signatures have been collected by the _Coordinator_, the signatures are inserted into the advice map (see also https://github.com/0xMiden/miden-client/issues/1077). At this point, calling `emit.AUTH_REQUEST` fetches the signature from the advice map.

So, while you are technically correct that the keystore (authenticator) will not hold the signatures, they are stored in the advice map, and are available to the execution host so that this event pushes the right signature to the advice stack.

Reply:
Thanks, Got it